zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java

```diff
@@ -341,41 +370,141 @@ IndexTemplates ensureIndexTemplates() {
+      .workerGroup(EventLoopGroups.newEventLoopGroup(
+        Flags.numCommonWorkers(), "armeria-common-worker", true), true)
+      .useHttp2Preface(false);
+    clientFactoryCustomizer().accept(builder);
+    return builder.build();
+  }
+
+  @Memoized // hosts resolution might imply a network call, and we might make a new client instance
+  public HttpClient httpClient() {
     List<String> hosts = hostsSupplier().get();
     if (hosts.isEmpty()) throw new IllegalArgumentException("no hosts configured");
-    OkHttpClient ok =
-        hosts.size() == 1
-            ? client()
-            : client()
-                .newBuilder()
-                .dns(PseudoAddressRecordSet.create(hosts, client().dns()))
-                .build();
-    ok.dispatcher().setMaxRequests(maxRequests());
-    ok.dispatcher().setMaxRequestsPerHost(maxRequests());
-    return new HttpCall.Factory(ok, HttpUrl.parse(hosts.get(0)));
+
+    List<URL> urls = hosts.stream()
+      .map(host -> {
+        try {
+          return new URL(host);
+        } catch (MalformedURLException e) {
+          throw new IllegalArgumentException("Invalid host: " + host, e);
+        }
+      })
+      .collect(Collectors.toList());
+
+    final EndpointGroup endpointGroup;
+    if (urls.size() == 1) {
+      URL url = urls.get(0);
+      if (isIpAddress(url.getHost())) {
+        endpointGroup = null;
+      } else {
+        DnsAddressEndpointGroupBuilder dnsEndpoint =
+          new DnsAddressEndpointGroupBuilder(url.getHost());
+        if (url.getPort() != -1) {
+          dnsEndpoint.port(url.getPort());
+        }
+        endpointGroup = dnsEndpoint.build();
+      }
+    } else {
+      List<EndpointGroup> endpointGroups = new ArrayList<>();
+      List<Endpoint> staticEndpoints = new ArrayList<>();
+      for (URL url : urls) {
+        if (isIpAddress(url.getHost())) {
+          staticEndpoints.add(Endpoint.parse(url.getAuthority()));
+        } else {
+          endpointGroups.add(url.getPort() == -1
+            ? DnsAddressEndpointGroup.of(url.getHost())
+            : DnsAddressEndpointGroup.of(url.getHost(), url.getPort()));
+        }
+      }
+
+      if (!staticEndpoints.isEmpty()) {
+        endpointGroups.add(new StaticEndpointGroup(staticEndpoints));
+      }
+
+      if (endpointGroups.size() == 1) {
+        endpointGroup = endpointGroups.get(0);
+      } else {
+        endpointGroup = new CompositeEndpointGroup(endpointGroups);
+      }
+    }
+
+    final String clientUrl;
+    if (endpointGroup != null) {
+      HttpHealthCheckedEndpointGroup healthChecked = new HttpHealthCheckedEndpointGroupBuilder(
+        endpointGroup, "/_cluster/health")
+        .clientFactory(clientFactory())
+        .build();
+      EndpointGroupRegistry.register(
+        "elasticsearch", healthChecked, EndpointSelectionStrategy.ROUND_ROBIN);
+      try {
+        healthChecked.awaitInitialEndpoints(30, TimeUnit.SECONDS);
```

Comment:
30 seconds is probably excessive btw. usually we allow services to come up later (we don't fail fast rather defer folks to using health check). For example it is a bug for the collector to break startup.